File: COA software/backend/app/models/audit.py
from datetime import datetime, timezone
from sqlalchemy import Column, String, DateTime, ForeignKey, JSON
from app.database.base import Base, generate_uuid, utc_now

class AuditLog(Base):
    __tablename__ = "audit_logs"

    id = Column(String(36), primary_key=True, default=generate_uuid)
    user_id = Column(String(36), ForeignKey("users.id", ondelete="SET NULL"), nullable=True, index=True)
    action = Column(String(64), nullable=False) # CREATE, UPDATE, DELETE, APPROVE, REJECT
    entity_type = Column(String(64), nullable=False, index=True)
    entity_id = Column(String(64), nullable=False, index=True)
    old_value = Column(JSON, nullable=True)
    new_value = Column(JSON, nullable=True)
    request_id = Column(String(64), nullable=True)
    created_at = Column(DateTime(timezone=True), default=utc_now, nullable=False, index=True)

File: COA software/frontend/src/types/audit.ts
export interface AuditLog {
  id: string
  user_id?: string | null
  username?: string | null
  action: string
  entity_type: string
  entity_id: string
  old_value?: Record<string, unknown> | null
  new_value?: Record<string, unknown> | null
  description?: string | null
  ip_address?: string | null
  created_at: string
}
